### DISCUSSION TITLE
Leistungsvariable Steuerung anderer Geräte, als Wallboxen

### DISCUSSION THREAD

**Discussion by ArminCoin** [camueller/SmartApplianceEnabler#218] [Installation und Betrieb]
Ich wollte mir für meinen Warmwasserbehälter jetzt einen Heizstab mit Stufen von 0,5 bis 3,5Kw/h holen und über den SAE steuern. Für den Heizstab wollte ich mir eine kleine REST API über die einen Raspberry Pi bauen, um die 6 Stufen vom SAE aus zu steuern. Diesen müsste ich dann als Wallbox im SAE anlegen, und die Stromstärke (Ampere) wie bei den Wallboxen an die API schicken, meine REST API würde dann den Rest erledigen und die 6 Stufen zuordnen. Oder gibt es da noch andere Möglichkeit einen Schalter einzubinden für Verschiedene Leistungsstufen, evtl. so das es nicht als Wallbox drin steht?

**Comment by camueller** [comment]
Ich denke, dass man sich bei dem Thema zuerst über die Hardware Gedanken machen sollte, dann über die Software.

Die einfachste Form eines regelbaren Heizstabes besteht darin, dass man jede Heizwendel einzeln schaltet. Ich habe selbst einen Heizstab mit 3x1,5kW, den ich allerdings nicht nutze. Zum Schalten würde man drei Solid State Relais verwenden, wobei jedes mit einen GPIO des Raspberry Pi verbunden ist.

Es gibt auch Heizstäbe, deren Leistung sich via PWM oder auch Phasenanschnittsteuerung steuern lässt. Dazu habe ich eine [Übersicht aus dem Jahr 2015](https://www.pv-magazine.de/archiv/bersicht-regelbare-heizstbe) gefunden - vielleicht findest Du aktuellere Übersichten/Produkte. Ohne anspruchsvolle Hardware-Bastelei kommt für den Raspberry Pi nur PWM in Frage. Allerdings produziert der Raspberry Pi ein PWM-Signal mit 5V und die Heizstäbe erwarten wohl meist eines mit 240V oder 400V. Zum Wandeln könnte man wohl einen [PWM Phasendimmer mit 5V Eingang](https://www.amazon.de/Krida-Phasendimmer-50-60HZ-Arduino-Raspberry/dp/B07CT8F9KX) verwenden.

Kommen wir zur Software: Es gab im [PV-Forum immer mal wieder Diskussionen zu PWM](https://www.photovoltaikforum.com/core/search-result/8036988/?highlight=PWM). Wie Du korrekt geschrieben hast unterstützt der SAE nur bei Wallboxen eine variable Leistungsaufnahme. Du kannst natürlich ein Script schreiben, das dem [SAE eine virtuelle Wallbox vorspielt](https://github.com/camueller/SmartApplianceEnabler/discussions/214), aber das würde ich als komplexe Krücke betrachten, weil die Wallbox-Steuerung mit SOC etc. unnötig komplex für die Ansteuerung von Heizstäben oder ähnlichen Geräten ist. Aber ich könnte mir durchaus vorstellen, dem SAE einen Schaltertyp für PWM und/oder einen für Multi-GPIO (entsprechend oben Abschnitt 1) hinzuzufügen, welchen dann dem Home Manager als Verbraucher mit variabler Leistungsaufnahme präsentiert werden. Da ich selbst keine Anwendung für dieses Feature habe, kann ich es nur entwickeln, wenn jemand das Testen übernimmt. Bevor ich den SAE anpasse müsste im Fall von PWM sichergestellt sein, dass es hardwareseitig überhaupt funktioniert. Das könnte man mit einem kleinen Script verifizieren, beispielsweise ähnlich dem in [diesem Tutorial](https://tutorials-raspberrypi.de/programmieren-lernen-raspberry-pi-gpio-pwm/).

**Comment by ArminCoin** [comment]
Also ich hatte vor diesen zu Kaufen
https://www.killus-technik.de/de/heizungszubehoer/elektro-heizstaebe/photovoltaik-vario-heizer-0-5-3-5-kw.html
Und dann hätte ich die einzelnen Schalter S1-S3 über drei 230v Relais über den Raspberry PI per GPIO gesteuert.(Über die 3 Schalter sind im 500w Schritten werte von 0,5 bis 3,5kw/h möglich. Phasenschnitt wäre natürlich auch interessant mit einem dafür kompatiblen Heizstab, damit wäre dann ja eine noch bessere Ausnutzung des Überschusses möglich.

**Comment by camueller** [comment]
Phasenanschnitt und PWM sind im Ergebnis gleich, aber mit dem Raspberry Pi geht nur PWM. 

Dafür müßte ich im SAE einen Multi-Switch mit Zuordnung zu SEMP-PowerLevels implementieren. Sollte aber kein Problem sein.
Initial würde ich das erstmal im Backend implemtieren, d.h Du musst manuell die Appliances.xml konfigurieren. Erst, wenn das alles passt, würde ich das Web-Frontend anpassen.

**Comment by NewUser1Z** [comment]
Guten Abend,
ich würde den selben Anwendungsfall mit PWM benötigen.
Zum Thema Hardware kann ich folgendes beitragen.
Meine bisherige Lösung habe ich hier gefunden: [Hier im Post #13 werden die Bauteile beschrieben.](https://forum-raspberrypi.de/forum/thread/35715-wirkleitstungscontroller/?postID=329672#post329672)
Ich kann berichten, dass die Bauteile gut funktionieren, leider ist die Software noch in der Entwicklung und läuft nicht so stabil.

Betreiben tue ich den 6KW (3x2KW) Heizstab aktuell auch nur mit einer Phase (2KW) so wie in der Beschreibung. Ob hier noch mehr möglich ist müsste ein Spezialist (Elektriker?) beurteilen.

Ich hoffe ich konnte zum Thema Hardware weiterhelfen.

**Reply by camueller** [reply]
Vielen Dank für den Link! Wenn ich das richtig verstehe, wird hier aber bei Überschreitung eines bestimmten PWM-Wertes ein Halbleiter-Relais ein- oder ausgeschaltet und der Heizstab selbst ist dann entweder ein- oder ausgeschaltet. Das kann man mit dem SAE auch ohne PWM jetzt schon mit einem Halbleiter-Relais. Ich sehe nicht, dass Du mit dieser Hardware den Heizstab mit variabler Leistungsaufnahme betreiben kannst, was ja eigentlich das Ziel ist. Oder irre ich mich da?

**Reply by NewUser1Z** [reply]
Fast Richtig, bei Überschreitung eines Schwellwertes (z.B. Leistung am Einspeisepunkt größer 6.800W) wird das PWM Signal hoch- bzw. bei Unterschreitung runtergefahren.
Richtig ist, hier wird ebenfalls ein Halbleiter-Relais ein- und Ausgeschaltet.
Dieses spezielle Halbleiterrelais kann man aber sehr schnell schalten und es reagiert sehr gut auf das PWM-Signal.
Ich konnte mit einem ShellyPlug eine stufenlose Regelung von 0W bis ca. 2.000W beobachten.

Meine Elektro Kenntnisse sind leider nicht so umfangreicht aber ich versuche es mal …
Dieses Halbleiterrelais lässt immer ganze Sinuswellen durch und schaltet bei Nulldurchgang. (Anders als Phasenanschnitt / Phasenabschnitt Geräte. Somit bessere Elektromagnetische Verträglichkeit.)
Somit kann man bei 50Hz auch mal nur 1Hz durchlassen. (Ich hoffe ich habe es soweit richtig erklärt.)

**Reply by camueller** [reply]
Verstehe ich Dich richtig, dass Du einen Shelly Plug verwenden willst, damit er einen Heizstab (der eigentlich nicht mit einem PWM-Signal umgehen kann) entsprechend des PWM-Signals und der PWM-Frequenz an- und auschaltet? Und diesen Aufbau hast Du bereits getestet, d.h. der Heizstab wurde heiss und die Leistungsaufnahme war zwischen 0 und 2 kW?

**Reply by camueller** [reply]
Ich habe einen Issue für die Details zur Implementierung der PWM-Unterstützung erstellt: https://github.com/camueller/SmartApplianceEnabler/issues/222

Aus meinen Modellbauzeiten hatte ich noch einen Servo, den ich gerade erfolgreich über PWM auf em Raspi steuern konnte! Grundsätzlich sollte also dem PWM-Support im SAE nichts im Wege stehen.

**Reply by NewUser1Z** [reply]
Ich versuche es nochmal mit der Beschreibung der Bauteile.
Raspberry: 	Erzeugt das PWM-Signal (feste Frequenz, variable Pulsbreite 0 bis 100%)
Optokoppler:	macht aus dem 5V PWM Signal ein 24V PMW Signal (24V Netzteil erforderlich)
Halbleiterrelais:	bekommt vom Optokoppler das Steuersignal (24V erforderlich) und gibt dem Heizstab die dementsprechende Leistung frei
Shelly Plug:	in meinem Aufbau nur zum Test und zur Kontrolle ob die Stufenlose Regelung funktioniert. Für die Funktion nicht notwendig.

Ich hoffe ich konnte weiterhelfen.
Bei Fragen stehe ich gerne zur Verfügung.


Und … ich finde es toll, dass hier über PMW nachgedacht und ggf. realisiert wird. Es gibt ja PWM zu 0-5V/0-10V Wandler. So bekommt der Raspberry (bzw. SMA) einen „Analogausgang“, das bringt mit Sicherheit wieder viele neue Möglichkeiten mit sich. **Weiter so!!!**

**Reply by premultiply** [reply]
Die Netzrückwirkungen von aufs Netz durchgeschalteter PWM dürften erheblich sein. Ich denke dass man das tunlichst nicht so machen sollte.
Klüger weil wesentlich störärmer sind hier Lösungen die sich z. B. über 3 separate Heizwendeln mit 3 Relais (gerne auch als Solid-State) in 8 Leistungsstufen schalten lassen.

**Reply by NewUser1Z** [reply]
@premultiply 
Grundsätzlich gebe ich dir Recht, aber das [Siemens Bauteil](https://mall.industry.siemens.com/mall/de/de/Catalog/Product/3RF2320-1CA04) ist denke ich genau für sowas ausgelegt. 
Vielleicht liest ja hier kurz ein Spezialist drüber und bestätigt das hoffentlich.

Generell muss es ja Beiteile auf dem Markt geben die für soetwas taugen. Fertigteile wie **AC ELWA-E** versprechen ja auch eine stufenlose Regelung. (Wäre natürlich interessant was die für Komponenten verbauen.)

**Reply by premultiply** [reply]
Letztendlich wird dies dort z. B. per Phasenanschnitt und vor allem entsprechenden Netzfiltern gelöst sein.
Insbesondere letztere würden bei der Selbstbaulösung fehlen.
Bei der Verwendung von SSR mit PWM im Selbstbau käme ggf. noch hinzu dass diese nur zu bestimmten Zeitpunkten zu und abschalten und somit eine Überlagerung von Netzfrequenz mit der PWM auftritt die alles andere als eine lineare Ansteuerung und eben erhebliche Netzrückwirkungen haben wird weil hier und da mal eine einzelne Halbwelle durchgeschaltet wird.

**Reply by NewUser1Z** [reply]
Ich habe den Eindruck du verstehst einiges von der Sache ...
Kann man bei dem Hinweis "Low Noise" davon ausgehen, dass hier Netzfilter schon mit drinnen sind bzw. nicht benötigt werden?

![EMV](https://user-images.githubusercontent.com/101476436/158076217-bab94e9c-2b42-4210-bd38-5dc4a3f3135b.JPG)

**Reply by ArminCoin** [reply]
Also Grundsätzlich wäre eine Lösung mit 3 Relais und so einem Heizstab
https://www.killus-technik.de/de/heizungszubehoer/elektro-heizstaebe/photovoltaik-vario-heizer-0-5-3-5-kw.html
aber besser oder?

**Reply by premultiply** [reply]
EMV ist leider ein ziemlich komplexes Thema das nicht einfach mit Standardbauteilen "aus der Bastelkiste" zu lösen ist sondern teils aufwendige Messreihen in speziellen Umgebunden, u. U. komplizierte Filterschaltungen und eben auch ein bestimmtes Schaltungsdesign erfordern kann. Insbesondere wenn es um Leistungselektronik geht wird es schnell komplex und aufwendig.
Käuflich erwerbliche Geräte und Systeme müssen nachweisen dass sie die entsprechenden Richtlinien zur EM-Abstrahlung und zu Netzrückwirkungen in allen Betriebszuständen einhalten.
Bei Selbstbau ist das kaum zu leisten aber dennoch unbedingt zu beachten da man sich sonst ggf. mehr, vielleicht zunächst oder selbst unbemerkte Störungen ins eigene Haus und (viel schlimmer) in das gesamte Netz holen kann.
Im dümmsten Fall steht irgendwann der Netzbetreiber und/oder der Funkstörungsmessdienst der BNetzA vor der Tür ...

Ohen das jetzt näher geschaut zu haben gehe ich davon aus, dass die oben genannten Standardfilter eben genau die durch den SSR-Betrieb alleine (ohne Betrachtung der angeschlossenen Last) verursachten Störungen durch den systemimmanenten leichten Phasenanschnitt eines SSR in jeder Halbwelle auf das notwendige Maß zu reduzieren.
Zur Vermeidung der mutmaßlich erheblichen Netzrückwirkungen bei variabler PWM-getakteter Ansteuerung einer großen Last durch ein SSR sind diese mit ziemlicher Sicherheit nicht geeignet.

**Reply by NewUser1Z** [reply]
@ArminCoin 
Was besser ist kann ich jetzt nicht beurteilen, liegt wahrscheinlich im Auge des Betrachters. (Geschmacksache)

Von der Auslegung der Größe kann ich dir von mir folgendes Berichten.
Ich habe mich für die 6KW Variante entschieden. Als Hauptheizung läuft bei mir eine Hackschnitzelheizung und im unwahrscheinlichen Fall, dass diese mal Ersatzteile braucht, kann ich mit dem Elektroheizstab überbrücken und sitze nicht in der Kälte. (gebt euch keine Mühe, das muss man nicht verstehen 😊)

**Comment by ArminCoin** [comment]
Ich denke bei mir wird es auf diesen Heizstab mit stufenlosem PWM Eingang hinauslaufen
https://www.ta.co.at/x2-energiemanagement/heizstab-3000-w-stufenlos-regelbar/
Ich suche noch nach einem der evtl. auch bis 6000 Watt geht, aber sonst wird dieser wohl auch gehen.

**Comment by camueller** [comment]
Ich habe einen Issue angelegt für die Details zum Implementierung des Multi-Switches: https://github.com/camueller/SmartApplianceEnabler/issues/223

**Comment by camueller** [comment]
Ich werde in den nächsten Tagen voraussichlich den PWM-Switch und den Multi-Switch implementieren für SAE 2.0.3. Den PWM-Switch kann ich mit einem Servo testen, aber außerdem kann ich testmäßig nicht allzuviel tun. Da kommt Ihr dann ins Spiel.

**Comment by BossoH** [comment]
Ich habe für diesen Zweck ein Gerät von Kemo verwendet.
Hier der [Leistungsregler](https://www.kemo-electronic.de/de/Licht-Ton/Effekte/Module/M028N-Leistungsregler-110-240-V-AC-4000-VA.php).
[Hier gibt es noch ein Modul](https://www.kemo-electronic.de/de/Wandler-Regler/Konverter/M150-DC-Puls-Converter.php), das aus einem PWM Signal des Raspberry den Baustein oben ansteuert.
Ich habe statt des 2. Moduls allerdings eine Relais Karte verwendet mit der ich Widerstände zu- bzw. abschalte. Ist viel billiger und für den Zweck braucht man keine stuffenlose Regelung.

**Comment by camueller** [comment]
Der PWM_Schalter ist implementiert - siehe Doku: https://github.com/camueller/SmartApplianceEnabler/blob/2.0/doc/PwmSwitch_DE.md

Als Nächstes implementiere ich den Multi-Switch und dann werden beide mit 2.0.4 releast ...

**Comment by camueller** [comment]
Der Stufenschalter ist ebenfalls implementiert - siehe Doku: https://github.com/camueller/SmartApplianceEnabler/blob/2.0/doc/LevelSwitch_DE.md

Es fehlt noch ein automatisierter Test für das Anlegen eine solchen Schalters und dann kann 2.0.4 releast werden ...

**Reply by BossoH** [reply]
Hallo Axel,
ich habe den Multischalter jetzt mal ausprobiert. Leider ohne Erfolg. Weder das manuelle Einschalten, noch das Zeitplan gesteuerte funktioniert. Die Ampel bleibt immer gelb.
Im Log sind mir die folgenden beiden Zeilen aufgefallen:
_DEBUG [Thread-4] d.a.s.c.Switch [Switch.java:94] F-*-000000000006-00: Starting Switch for GPIO 22
WARN [Thread-4] d.a.s.c.Switch [GpioControllable.java:90] F-*-000000000006-00: Configured for 22, but GPIO access disabled._ 

Das wiederholt sich für alle 4 GPIO Schalter, die ich konfiguriert habe.
Hast du eine Idee, ob das auf die Ursache deutet und wie ich das Problem löse?
Danke.

**Reply by camueller** [reply]
Ohne vollständiges Log kann ich nicht helfen.

**Reply by BossoH** [reply]
[Hier ist ein aktuelles Log.](https://github.com/camueller/SmartApplianceEnabler/files/9019466/rolling-2022-06-30.log)
 Kurz vor Ende ist ein händisches Einschalten. Automatisch hätte er das die ganze Zeit immer wieder gekonnt.
Stecke aber bitte nicht zu viel Mühe hinein. Wenn es nicht funktioniert ist es auch nicht schlimm, dann wechsel ich wieder zu meiner bisherigen Lösung zurück. Wollte die Funktion nur aus Neugierde mal ausprobieren.

**Reply by camueller** [reply]
Mir fallen im Log vor zunächst folgende Zeilen auf, die da nicht stehen sollten:
```
2022-06-30 10:12:43,246 DEBUG [Thread-4] d.a.s.c.Switch [Switch.java:94] F-*-000000000006-00: Starting Switch for GPIO 22
2022-06-30 10:12:43,248 WARN [Thread-4] d.a.s.c.Switch [GpioControllable.java:90] F-*-000000000006-00: Configured for 22, but GPIO access disabled.
2022-06-30 10:12:43,249 DEBUG [Thread-4] d.a.s.c.Switch [Switch.java:94] F-*-000000000006-00: Starting Switch for GPIO 23
2022-06-30 10:12:43,250 WARN [Thread-4] d.a.s.c.Switch [GpioControllable.java:90] F-*-000000000006-00: Configured for 23, but GPIO access disabled.
2022-06-30 10:12:43,251 DEBUG [Thread-4] d.a.s.c.Switch [Switch.java:94] F-*-000000000006-00: Starting Switch for GPIO 24
2022-06-30 10:12:43,252 WARN [Thread-4] d.a.s.c.Switch [GpioControllable.java:90] F-*-000000000006-00: Configured for 24, but GPIO access disabled.
2022-06-30 10:12:43,253 DEBUG [Thread-4] d.a.s.c.Switch [Switch.java:94] F-*-000000000006-00: Starting Switch for GPIO 25
2022-06-30 10:12:43,254 WARN [Thread-4] d.a.s.c.Switch [GpioControllable.java:90] F-*-000000000006-00: Configured for 25, but GPIO access disabled.
```

Stammt das Log von einem Raspberry Pi?
Lief auf diesem der [pigpiod](https://github.com/camueller/SmartApplianceEnabler/blob/2.0/doc/ManualInstallation_DE.md#pigpiod-installieren) - er muss laufen für den GPIO-Zugriff!

**Reply by BossoH** [reply]
Ja und ja, siehe Screenshot:
![image](https://user-images.githubusercontent.com/89273714/176677418-c3e2e469-9c38-4400-bda2-e87d5304e34b.png)

**Reply by camueller** [reply]
Da ist da noch ein Fehler drin, den ich für die nächste 2.0-Version beheben werde. Dann kannst Du nochmal testen :-)

**Reply by camueller** [reply]
Es waren sogar 2 Fehler drin :-) - sind jetzt gefixt für die 2.0.5, die demnächst release wird. Bitte dann nochmal testen ...

**Reply by camueller** [reply]
2.0.5 ist releast - Du kannst den Stufenschalter testen :-)

**Reply by BossoH** [reply]
Sorry, hat ein bisschen gedauert, bis ich die Zeit gefunden habe.
Der vorherige Fehler ist nun weg, funktionieren tut es aber leider trotzdem nicht.
[Hier ein Logfile.](https://github.com/camueller/SmartApplianceEnabler/files/9228793/rolling-2022-07-31.log) Es schaltet weder beim Klick auf die Ampel noch automatisch ein.

**Reply by camueller** [reply]
Kann mir das Log-File erst nach meinem Urlaub ansehen ...

**Reply by BossoH** [reply]
Ich wünsche dir einen schönen und erholsamen Urlaub.

**Reply by Lausbua** [reply]
@BossoH 
Funktioniert dein Stufenschalter mittlerweile oder ist das Problem noch aktuell?

**Reply by BossoH** [reply]
Der Stufenschalter funktioniert bei mir nicht.
Ich verwende nach wie vor meine "Pythonwallbox".

**Comment by BossoH** [comment]
Ich habe es nun auch mit Version 2.0.6 probiert. Das manuelle Schalten per Ampel funktioniert jetzt. Automatisch schaltet es aber nach wie vor nicht.
[Hier ein Log.](https://github.com/camueller/SmartApplianceEnabler/files/9882263/rolling-2022-10-27.log)

**Reply by camueller** [reply]
Dein Log-File enthält aber eine Vielzahl von Einschaltbefehlen mit verschiedenen Leistungsvorgaben, z.B.:
```
2022-10-27 13:03:04,530 DEBUG [http-nio-192.168.0.9-8080-exec-3] d.a.s.s.w.SempController [SempController.java:300] F-########-000000000006-00: Received control request: on=true, recommendedPowerConsumption=1000W
2022-10-27 13:03:04,538 DEBUG [http-nio-192.168.0.9-8080-exec-3] d.a.s.a.Appliance [Appliance.java:433] F-########-000000000006-00: Setting appliance state to ON
2022-10-27 13:03:04,543 INFO [MQTT Call: F-########-000000000006-00-LevelSwitch-0] d.a.s.c.LevelSwitch [LevelSwitch.java:214] F-########-000000000006-00: Setting power to 1000W
2022-10-27 13:03:04,543 DEBUG [MQTT Call: F-########-000000000006-00-LevelSwitch-0] d.a.s.c.LevelSwitch [LevelSwitch.java:206] F-########-000000000006-00: Setting wrapped control switch 1 to on
2022-10-27 13:03:04,544 DEBUG [MQTT Call: F-########-000000000006-00-LevelSwitch-0] d.a.s.c.LevelSwitch [LevelSwitch.java:206] F-########-000000000006-00: Setting wrapped control switch 2 to on
2022-10-27 13:03:04,545 DEBUG [MQTT Call: F-########-000000000006-00-LevelSwitch-0] d.a.s.c.LevelSwitch [LevelSwitch.java:206] F-########-000000000006-00: Setting wrapped control switch 3 to off
2022-10-27 13:03:04,547 DEBUG [MQTT Call: F-########-000000000006-00-LevelSwitch-0] d.a.s.c.LevelSwitch [LevelSwitch.java:206] F-########-000000000006-00: Setting wrapped control switch 4 to off
2022-10-27 13:03:04,550 INFO [MQTT Call: F-########-000000000006-00-Switch-0] d.a.s.c.Switch [Switch.java:144] F-########-000000000006-00: Switching on GPIO 22
2022-10-27 13:03:04,601 INFO [MQTT Call: F-########-000000000006-00-Switch-2] d.a.s.c.Switch [Switch.java:144] F-########-000000000006-00: Switching off GPIO 24
2022-10-27 13:03:04,601 INFO [MQTT Call: F-########-000000000006-00-Switch-1] d.a.s.c.Switch [Switch.java:144] F-########-000000000006-00: Switching on GPIO 23
2022-10-27 13:03:04,601 INFO [MQTT Call: F-########-000000000006-00-Switch-3] d.a.s.c.Switch [Switch.java:144] F-########-000000000006-00: Switching off GPIO 25
```
